build_tools/autogen_ltc_backend.py

```diff
@@ -0,0 +1,290 @@
+        "to.dtype",  # Error: TODO add support for type BaseType(name=<BaseTy.MemoryFormat: 12>)
+        "to.other",  # Error: TODO add support for type BaseType(name=<BaseTy.MemoryFormat: 12>)
+        "uniform_",  # Error: TODO add support for type BaseType(name=<BaseTy.MemoryFormat: 12>)
+        "zeros",  # Error: Code below assumes there is at least one tensor arg
+        "zeros_like",  # Error: TODO add support for type BaseType(name=<BaseTy.MemoryFormat: 12>)
+    }
+
+    # Additional ops which autogen is supported for but don't compile yet
+    blacklist |= {"item", "size", "where"}
+
+    # List of supported ops that we don't want to do the full codegen for
+    # primarily view ops
+    supported = {
+        "expand",
+        # "native_batch_norm_backward",
+        "native_batch_norm",
+        "permute",
+        "repeat",
+        "squeeze",
+        "t",
+        "unsqueeze",
+        "view",
+    }
+
+    if which("rg") is not None:  # use ripgrep if available as its much faster
+        cmd = ["rg", "-o", "-N", r"aten::[0-9a-zA-Z_\.]+"]
+    else:
+        cmd = ["grep", "-o", r"aten::[0-9a-zA-Z_\.]\+"]
+
+    output = (
+        subprocess.check_output(
+            cmd + [str(aten_ops_file)],
+            encoding="utf-8",
+        )
+        .strip()
+        .split(os.linesep)
+    )
+
+    # process ops list
+    ops = []
+    supported_ops = []
+    skipped = []
+
+    for op in output:
+        op = op[6:]
+        opname = op.split(".")[0]
+
+        if opname in blacklist or op in blacklist:
+            continue
+
+        if opname in supported:
+            supported_ops.append(op)
+            continue
+
+        if op not in aten_funcs:
+            skipped.append(op)
+            continue
+
+        ops.append(op)
+
+    opnames = sorted(set(ops))
+
+    with out_file.open("w") as f:
+        yaml.dump(
+            {
+                "backend": "Lazy",
+                "cpp_namespace": "torch_lazy_tensors",
+                "full_codegen": opnames,
+                "supported": sorted(supported_ops),
+            },
+            f,
+            default_flow_style=False,
+        )
+        f.write(
+            dedent(
+                """
+
+                # Skipped ops (supported by Torch-MLIR but no equivalent native function)
+                """
+            )
+            + os.linesep.join(f"#  - {op}" for op in sorted(skipped))
+        )
+
+    return parsed_yaml, grouped_native_functions
+
+
+@dataclass(frozen=True)
+class MlirLazyIr(codegen.gen_lazy_tensor.dest.LazyIR):
+    lowering_function_type: str = "torch::lazy::MlirFunction"
+    lowering_context_type: str = "torch::lazy::MlirLoweringContext*"
+    lowering_return_type: str = "torch::lazy::MlirOpVector"
+
+    def lowering_body(self, f):
+        func = (
+            f.functional.func if isinstance(f, NativeFunctionsGroup) else f.func
+        )
+        schema = LazyIrSchema(func)
+
+        return f"""
+        UNIMPLEMENTED_ERROR(
```

Comment:
i'm curious, how do you plan to code-generate the lowerings to MLIR?  i guess you have the aten-mlir dialect and can piggy-back off that?

Reply:
The design for that is still up in the air. However, I was thinking that I could do something similar to what you did here
https://github.com/pytorch/pytorch/blob/d9896b8b4f198fff43ff8003553c13a26381c7ce/tools/codegen/dest/lazy_ts_lowering.py#L37
And add a generic `LowerMlirBuiltin` function that handles lowering to MLIR. There is a Torch-MLIR ATen dialect that will be used here, yes.

Reply:
yea im less concerned with how you call or generate some function, more curious where the source of truth for the lowering is maintained (And if you had some way of actually generating the lowering as opposed to hand-implementing the lowering)

Reply:
I would definitely prefer if we went through the JIR IR path rather than directly creating Lazy -> MLIR. That would let us more easily share op decompositions and other infra built at the JIT IR level.

Ideally upstream would allow us to share all the ts_backend code and just need to implement the runtime bits and TS -> MLIR bits, but not LTC Node -> TS Node.

cc @Chillee

Reply:
cc @eellison

Reply:
> I would definitely prefer if we went through the JIR IR path rather than directly creating Lazy -> MLIR.

I think(?) this makes sense? Like, Torch-MLIR has spent the most time investing in a Torchscript => MLIR lowering, and LazyTensor has spent a lot of effort in a LTC => Torchscript lowering. Is there any reason these couldn't be (relatively) trivially combined together, instead of adding a whole new lowering path for MLIR?

I don't know the other design restrictions here, but that makes sense to me.

Reply:
Yea, I think aside from one issue i'm aware of, which is that we haven't landed the entire TS backend yet and it's tough for the MLIR team to take a c++ build dep on our code until we do land it all and they can pick it up in nightlies, this should be the best plan and was our intent.  I'm trying to push for landing it sooner rather than later, but its at least a month out still.